### ISSUE TITLE
Add DIrectx11/12 support and port to UWP for Xbox One dev Mode 

### ISSUE THREAD

**Issue by danprice142** [danoon2/Boxedwine#5]
As this is openGL it wont work on xbox one and above could direct x be added and then Boxedwine turned into a uwp/appx to allow it to work on xbox one?

**Comment by 39p** [comment]
Seems interesting and I would like to see this but as you said it can only run directx and not OpenGL but performance could be good looking at how emulation works on dev mode especially on the series consoles

**Comment by danoon2** [comment]
Boxedwine runs Wine by proving an emulated CPU and emulate Linux Kernel.  So Wine thinks its running on a 32-bit Linux system.  Wine translates DirectX to OpenGL, then Boxedwine intercepts and marshals the OpenGL calls from the emulated program to the host computer.  To do this I replaced the Linux version of libGL and created 1400 functions (lots of GL extensions) in there to do the forwarding for each one.

I think it would be possible to replace the Wine version of the DirectX libraries and forward those calls to the host, but this sounds like a huge project.

**Comment by danprice142** [comment]
I posted this on a RetroArch Discord channel and it got over 50 likes in 5 mins before we all realized it wouldn`t work how it is. I believe that even though this is a lot of work i feel like it could have a serious demand also if it works for xbox it will work on windows as a store app. The Xbox Support DirectX11/12 and has ANGLE too if that helps?

**Comment by Ryan14411** [comment]
I too would like this i so hope you do end up making the changes needed to get this working on the Xbox. keeping my fingers crossed. /
 And Thanks for Boxedwine im loving it! also would i be able to get red alert 2 working on boxedwine?

**Comment by Ryan14411** [comment]
https://www.theverge.com/2021/4/20/22393655/microsoft-windows-app-store-overhaul-apps-games-rumor

https://www.youtube.com/watch?v=f8_nvJzuaSU

https://devblogs.microsoft.com/commandline/the-initial-preview-of-gui-app-support-is-now-available-for-the-windows-subsystem-for-linux-2/ 

It uses Microsoft's own OpenGLOn12, by the way, which is an open source implemention of OpenGL, not proxying so i dont know if this will help making it a uwp for xbox but i think it will

**Comment by danprice142** [comment]
I found something interesting and may be pretty useful to you and may possibly work to bring boxed wine to UWP https://github.com/gfx-rs/portability https://www.khronos.org/blog/khronos-announces-the-vulkan-portability-initiative

**Comment by danoon2** [comment]
@danprice142   This one also looks interesting, and I was able to get it to work on Windows 10.

https://docs.mesa3d.org/drivers/d3d12.html

It converts OpenGL to Direct 3D 12.

**Comment by danprice142** [comment]
Nice, do you think these layers would dramatically cut out most of the work to bring Boxed Wine to UWP?

**Comment by danoon2** [comment]
I looked up more info on XBox One and it seems like XBox doesn't support dynamic code generation, at least according to this article

https://unity.com/how-to/develop-console-video-games-unity

In theory, Boxedwine could ahead of time recompile a particular game, but currently I don't have this feature.

**Comment by danprice142** [comment]
Im almost 100% sure it does, im sure retroarch in dev mode uses it, i just asked another dev

**Comment by danprice142** [comment]
https://github.com/bminor/SDL/blob/master/docs/README-winrt.md this could maybe make things easier

**Comment by toastmod** [comment]
What I've gathered from this thread is that wine's DX to OpenGL translation layer should probably be modified to call DX directly in UWP, I think that is the most optimal thing. Your other option is DX->OpenGL->ANGLE->DX which is super inefficient but would get the job done. I personally haven't had luck with ANGLE on Xbox but I didn't try overly hard either. So I think taking a look at the wine source and figuring out where that translator is might be the best solution here.

**Comment by toastmod** [comment]
So far I can see [something interesting here](https://github.com/wine-mirror/wine/blob/master/dlls/d3d12/d3d12_main.c), this seems to be where the D3D12 dll is written directly into Vulkan, however looking at the code for [D3D11](https://github.com/wine-mirror/wine/blob/master/dlls/d3d11/d3d11_main.c) the process seems to be a lot different. This is probably because [D3D12 is supposedly very similar to Vulkan](https://www.reddit.com/r/vulkan/comments/grx2a9/comment/fs34hmx/?utm_source=share&utm_medium=web2x&context=3). I'd like to learn more about what you went through to create your OpenGL passthrough driver for wine. 

So basically in order to do Wine <-> UWP I would make a libD3D12 for linux built the same way your marshalling libGL was built. Then I would implement this libD3D into the source code I linked above through thunks, like how [Vulkan is marshalled to the linux Vulkan library somewhere around here.](https://github.com/wine-mirror/wine/tree/master/dlls/winevulkan).

I don't know how thunking entirely works yet but I would like to look into it if I have the time. So there's some stuff to think about I guess.

**Comment by danoon2** [comment]
@toastmod   I'm currently in the process of implementing vulkan pass through

The first step is to get the call out of the emulator and into boxedwine with a stub/proxy, like what you mentioned with my libGL implementation

https://github.com/danoon2/Boxedwine/tree/james/vulkan/tools/vulkan

Then you need to intercept the call in Boxedwine like I do with OpenGL

https://github.com/danoon2/Boxedwine/blob/e7e5a363c145e23a3c1b6a2d717b7c59750c8aaa/source/emulation/cpu/x64/x64Ops.cpp#L1153
https://github.com/danoon2/Boxedwine/blob/e7e5a363c145e23a3c1b6a2d717b7c59750c8aaa/source/emulation/cpu/normal/normal_other.h#L203

You could try using int 3b for D3D pass through, each CPU emulation will need some code to forward that instruction

The decoder will also need a small update
https://github.com/danoon2/Boxedwine/blob/e7e5a363c145e23a3c1b6a2d717b7c59750c8aaa/source/emulation/cpu/decoder.cpp#L3644

The hard part is the marshaling.  Boxedwine only emulates 32-bit programs, so if you want to run on a 64-bit platform, then your API you want to pass through will will need some work.  Some data types might be a different size on the 64-bit platform.

See 

https://github.com/danoon2/Boxedwine/blob/master/source/opengl/glMarshal.h

For ideas on how to marshal on 32-bit and 64-bit platforms.

Simple structures that contains no pointers and all the members are the same size just need to get the real address from the emulated address (getPhysicalAddress)

The hardest part I haven't fully solved, is if the API returns back a pointer that points to read/write memory, like mapping memory functions.

**Comment by rodrigoandrigo** [comment]
https://devblogs.microsoft.com/directx/in-the-works-opencl-and-opengl-mapping-layers-to-directx/ this may help

**Comment by Pedro1234-code** [comment]
guys is there something new here, like, it is still open?

**Comment by toastmod** [comment]
@Pedro1234-code it's been a while, have had other things to do but will probably come back to this someday soon.

**Comment by danprice142** [comment]
Its been a while since I comented on this but I found something interesting https://github.com/aerisarn/mesa-uwp https://github.com/aerisarn/glxgears-uwp I have been talking to the dev and I know he has successfully got Mesa running in the Duckstation UWP emulator using OpenGL sucessfully on Xbox, as well as another emulator I can mention just yet, but that is two successfull tests

**Comment by aerisarn** [comment]
I was able to run this over Mesa UWP port with GL3 imgui. The main issue is that UWP doesn't support allocation on READ_WRITE_EXECUTE. With its codegen permission you can only call VirtualProtect over a page and put it in READ_EXECUTE mode, so the generator would have to fill a page and protect it before executing. From a quick glance, the memory allocation inside the binary translation would not support that as it's not exactly organized as pages

**Comment by danoon2** [comment]
@aerisarn That is interesting.  You are correct.  The current binary translator wants to generate the code in small increments.

I can only think of 2 solutions for this

1) change the code so that it will generate as much code as possible to fill a page when code is missing.  So for all branches where it jumps to a known location, just keep translating.

2) Allow the code to fall back to the normal emulation (no binary translation).  I might get to something like this later this year.  It's hard for me to know for sure.  One goal of mine has been to improve the web (WASM/emscripten) version of Boxedwine by introducing a WASM JIT.  But there is probably a limit to how many WASM pieces of code I can generate at run time, so I would have to batch them up.  I was thinking of making the normal CPU emulation multi-threaded then allow the JIT to run on that.

For performance, #1 would probably be better.  The downside is if a page can't be filled then it would waste memory.

**Comment by danprice142** [comment]
@danoon2 Thought you would like to know that Boxedwine is up and running on Xbox consoles, a UWP port has been completed by a developer called moonpower it can be found on xboxdev.store if you wanted to check it out

**Comment by danoon2** [comment]
@danprice142  I'm curious why it's using WASM.  Even if we can't use JIT on xbox, the c++ non JIT code should be at least 2x faster than WASM.

**Comment by momo-AUX1** [comment]
I did it on WASM because I know JS enough to maybe modify emcc and the resulting build if the need arises. As the others said, UWP has some strict limits that WASM luckily doesn't have so it was more "straightforward" all the modifications needed were to add a front for the url params and modify the base64 data extraction to work under UWP.

**Comment by ludaka** [comment]
> I did it on WASM because I know JS enough to maybe modify emcc and the resulting build if the need arises. As the others said, UWP has some strict limits that WASM luckily doesn't have so it was more "straightforward" all the modifications needed were to add a front for the url params and modify the base64 data extraction to work under UWP.

hey where can check your workor or it's closed srouce for now ? it will be cool if official Boxedwine to added it

**Comment by worleydl** [comment]
Here's a hacked up native version but I don't have any way to control it yet so largely untested: https://github.com/worleydl/Boxedwine-uwp

I didn't run into the allocation issues with this configuration yet.  I was trying to get setup to take a look but maybe somehow I bypassed the issue with how the project is configured.

**Comment by worleydl** [comment]
The GL pipeline is currently broken in the above fork, this is what gets spammed to the console I haven't had much luck debugging the root cause yet:
```
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
 libGL glXGetProcessAddressARB: glXChooseVisual result=1200DDF1
```

That result code is the same as when it's working correctly on Windows but we're getting stuck in a loop instead of moving onto the other functions.

**Comment by danoon2** [comment]
@worleydl Are you running via the command line?  If so, please be aware as of a few weeks ago I committed a change that broke the old file systems.  I now emulate x11 instead of using a custom winex11.drv replacement.  I can't tell from the log above, but maybe you need to update to the newer file system.  You can download it via the boxedwine UI or here is a direct link

http://boxedwine.org/v2/2/TinyCore15Wine6.0.zip

Also please be aware, with a new file system, you will also need to start fresh with a local directory for the root folder.

**Comment by worleydl** [comment]
I'm using the imgui UI to setup containers and experiment but after they boot I have no way of controlling the kb/mouse.  The main sketchy thing I'm doing is using a sync'd up windows build to prep containers then I copy those over to my UWP environment to try and get it to issue a "int99 callback" so I can continue working on the OGL init issues.

Input can eventually work.  We need a "virtual mouse" cursor added since native cursors don't work on xbox UWP then a shortcut to show the virtual keyboard and make sure it's passing thru.  I may go ahead and look at that to confirm I'm not breaking something by copying containers around.

**Comment by danoon2** [comment]
@worleydl PlatformOpenGL::init() is new code that will query Windows for Opengl info, it does this by creating and destroying multiple windows.  One thing to try might be to comment out the call to queryOpenGL in that function.

**Comment by worleydl** [comment]
I had done a little work to get queryOpenGL working with the single core window: https://github.com/worleydl/Boxedwine-uwp/blob/uwp-compat/platform/windows/platformOpenGL.cpp#L104

Something like the Full Tilt Pinball demo which doesn't require OpenGL will load and execute with sound.  But something like WinQuake or glxgears gets stuck in a loop and never seems to trigger the callbacks that executes code inside of sdlgl.cpp.  sdlgl.cpp still needs work in my fork but I'll need to figure out what's up with the callbacks never triggering.  I'm guessing maybe it's upset about the pixel format on the linux side and gets stuck trying to switch it, I think mesa-uwp keeps it locked to a single one at the moment.

**Comment by worleydl** [comment]
When I look at the wine code it should be loading other functions then failing further down the line.  I do see that functions are loaded inside a do/while(0) block and I have a bad feeling we're stuck in that loop for some reason.

**Comment by danoon2** [comment]
@worleydl What configuration are you using?

64-bit build?
BOXEDWINE_MULTI_THREADED defined?

Can you paste all the preprocessor defines you are using into this issue?

That would be weird to be stuck in a while (0) loop, I assume that gets compiled out.  But from the emulation side of things, its not hard to imagine the emulated instruction pointer not being incremented probably.

Are you testing on XBox hardware?  Does winquake run with a similarly configured Boxedwine on Windows?

I just tested motorhead (3dfx with OpenGL glide emulation) on win32 and win64 builds with the normal cpu (no binary translator) in single and multi-threaded mode.  Multi-threaded worked for me, single-threaded stopped as soon as it tried to move to 3dfx (opengl).  But I don't get the same error as you, I got

0108:fixme:d3d:wined3d_context_gl_check_fbo_status FBO status GL_FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT (0x8cd7).

Have you tried something simpler? Like http://boxedwine.org/demos/wglgears.zip

**Comment by worleydl** [comment]
Here's my current preproc list:
```
BOXEDWINE_OPENGL_SDL
BOXEDWINE_OPENGL_OSMESA
__WINDOWS_WASAPI__
BOXEDWINE_UI_LAUNCH_IN_PROCESS
BOXEDWINE_BINARY_TRANSLATOR
BOXEDWINE_HIGHDPI
GLEW_STATIC
IMGUI_IMPL_OPENGL_LOADER_GLEW
BOXEDWINE_X64
BOXEDWINE_MULTI_THREADED
BOXEDWINE_MIDI
BOXEDWINE_RECORDER_X
BOXEDWINE_64
BOXEDWINE_ZLIB
GLH=<sdl_opengl.h>
BOXEDWINE_MSVC
_NO_DEBUG_HEAP=0
WIN32
_DEBUG
_CONSOLE
```

I've tried variations with the BT/X64/Multithreading disabled with no luck.

In terms of my testing I'll find something that works on the latest windows build then copy those containers to the xbox for evaluation.  I've tried a different build of glgears which triggered the same behavior when it came to running code in libGL(gl.c).  From what I can gather it's never making it to the point of switching resolutions, from the above trace it never makes it past get_pixel_formats which triggers the init of OpenGL.  I will try your version later today to confirm.

Also I tried switching the windows build to /MD & /DYNAMICBASE and it still works so I don't know if those settings on the UWP project have factored into anything.

**Comment by danoon2** [comment]
@worleydl 

I didn't realize the binary translator would work on the xbox since running generated code can cause security issues.  On Mac, it requires special permissions in the project, for example.

As for debugging.  Have you tried WINEDEBUG channels?  You can set it up from the command line using -env "WINEDEBUG=+wgl", but if you are running from Visual Studio, I find it easier to just add it to the code by uncommenting

https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/sdl/startupArgs.cpp#L384

This isn't complete, but x11 logging might show something, you can set this to true

https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/x11/xserver.h#L86

syscall logging will spam a lot.  Since your bug is around querying opengl, boxedwine is already pretty far into starting up wine, so this might not be too useful

https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/kernel/syscall.cpp#L61

mmap syscalls all responsible for a lot of the spamming, so you can try setting the mask to ~SYSCALL_MEMORY

I really don't expect logging syscalls to tell you anything, but when I get desperate sometimes I try.

For the most complicated and time consuming logging, you can try to log the emulated instructions.  Sometimes this can cause bugs, so making it as surgical as possible can help

uncomment https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/emulation/cpu/x64/x64Asm.cpp#L4468

If you know you are only interested in the main process, you can wrap that in an if statement

if (cpu->thread->process->id == 0xa) {
    this->logOp(this->ip);
}

Then to actually log it, you have to open a log file for the thread.  You can log the entire thread's lifetime, or even all threads.  Since you are far into running wine, this could be gigabytes of logs, so it might help to find an if condition to enable it on a particular thread.

for all threads or 1 thread logging, see

https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/kernel/kthread.cpp#L122

BString tmp = BString::valueOf(id);
tmp += ".txt";
if (id==0xb)
    this->cpu->logFile.createNew(tmp);

This is where syscall logging can help or maybe you know from setting break points, to find out the process/thread id's where the loop is happening.

You can also use the above code anywhere in boxedwine, for example, if you can narrow down where the bug happens, like after a particular syscall or if you know the ip address you can enable the logging there

void X64Asm::translateInstruction() {
    KMemoryData* mem = getMemData(cpu->memory);
    this->startOfOpIp = this->ip;
    if (this->ip == 0xD1234567 && !cpu->logFile.isOpen()) {
        cpu->logFile.createNew(B("test.txt"));
    }

This all gets into some pretty crazy debugging.  Logging the emulated ops can really help, but only if its possible to narrow it down first.

**Comment by danoon2** [comment]
@worleydl On more thought, you mentioned building your own wine, are you testing with yours or mine?  If you are using yours, definitely make sure it works on Windows.  Building wine is tricky.  Currently I use 32-bit Debian 11 in a VM.  I tried Debian 12, but that led to some bugs, probably because of my kernel emulation.

**Comment by worleydl** [comment]
Yea it still runs on windows.  I made a i386 deb10 Dockerfile for doing Linux builds in case that's handy for anything.

Oddly I can't get the wgl WINEDEBUG channel to output anything but d3d/all work as expected.  I was able to confirm my changes loaded in because I rearranged the order of GL function loading.

**Comment by worleydl** [comment]
Thanks for the assist danoon2, the issue was isolated to devtty writes, if I always return the len the app proceeds to the sdlgl code (there's probably a better fix but at least it's been pinpointed now):

https://github.com/danoon2/Boxedwine/blob/master/source/kernel/devs/devtty.cpp#L139

**Comment by danoon2** [comment]
@worleydl That is really good news.  Do you think the xbox limits stdout?  Maybe there is an issue of writing to handle "1".  Seems like I probably should have used a constant, like "return fwrite(buffer, len, 1, stdout);"

**Comment by worleydl** [comment]
I haven't quite figured out stdout on UWP yet, we can get debug output using OutputDebugString but regular stdout goes into the void without some sorta special redirection.

**Comment by worleydl** [comment]
Finally got some gibberish to spit out with glgears.  It'll be a while before I figure out managing all the contexts on UWP within a single window but I welcome suggestions.  

UWP opengl stuff didn't like running multithreaded but if I wrapped the int99 callbacks to run on the ui thread then the plumbing works.  It should render correctly if we isolate the opengl context and only render that.  My other idea is making a very cheap windowing system that uses framebuffers but not sure if that's actually possible.

**Comment by worleydl** [comment]
![Screenshot_2024-12-15_21-55-32](https://github.com/user-attachments/assets/6fa53a00-e9be-456a-a3af-1cd606770c38)

😎

**Comment by danoon2** [comment]
Mac is the same way, all opengl has to be on the main thread

https://github.com/worleydl/Boxedwine-uwp/blob/36ff8fcfd15f1b5a40e569e134f24b92b8623f51/source/opengl/sdl/sdlgl.cpp#L265

**Comment by worleydl** [comment]
I modified SDL-UWP to allow for creating multiple windows under the assumption only one is shown at a time.  This allows for winquake and a few other titles to boot.  The motorhead demo does actually run but gets corrupted gfx once it tries to do 3D so there's some stuff to investigate around the glide wrapper still.

This line was causing me issues when using a multithreaded configuration. 
 If I comment it out I get speeds similar to running on desktop. with BT/MT/X64: https://github.com/danoon2/Boxedwine/blob/master/source/sdl/multiThreaded/threadedMainloop.cpp#L91

**Comment by worleydl** [comment]
I patched SDL & mesa to be happier with threads which resolved the WaitForEvent bottleneck issue.

**Comment by ludaka** [comment]
hey do you think we can remove cpu emulation when processor native rn at x64 ?

**Comment by danoon2** [comment]
@ludaka  To make Wine run in Boxedwine, I went down the path of emulating Linux.  That requires emulating processes and memory address spaces.   Without re-creating a full VM like VirtualBox and using a kernel driver to get access to changing x86 segments and the MMU, it wouldn't be possible to emulate Linux without a CPU emulator.  It might be interesting to make a kernel driver some day.  But for the last few years I have mainly been focused on getting newer versions of Wine to run and getting parity between Windows x64 systems and Arm Mac computers.

Basically as 1 person, my TODO list grows faster than what I can actually do 😄
